SpatialGDK/Source/Private/Interop/GlobalStateManager.cpp

```diff
@@ -93,97 +93,149 @@ void UGlobalStateManager::ApplyAcceptingPlayersUpdate(bool bAcceptingPlayersUpda
+	Channel = Cast<USpatialActorChannel>(Connection->CreateChannel(CHTYPE_Actor, 1));
 
-		NetDriver->PackageMap->ResolveEntityActor(SingletonActor, SingletonEntityId, improbable::CreateOffsetMapFromActor(SingletonActor, Info));
+	SingletonActor->Role = ROLE_SimulatedProxy;
+	SingletonActor->RemoteRole = ROLE_Authority;
 
-		UE_LOG(LogGlobalStateManager, Log, TEXT("Linked Singleton Actor %s with id %d"), *SingletonActor->GetClass()->GetName(), SingletonEntityId);
-	}
+	// Add to entity registry
+	// This indirectly causes SetChannelActor to not create a new entity for this actor
+	NetDriver->GetEntityRegistry()->AddToRegistry(SingletonEntityId, SingletonActor);
+
+	Channel->SetChannelActor(SingletonActor);
+
+	// Since the entity already exists, we have to handle setting up the PackageMap properly for this Actor
+	FClassInfo* Info = NetDriver->TypebindingManager->FindClassInfoByClass(SingletonActor->GetClass());
+	check(Info);
+
+	NetDriver->PackageMap->ResolveEntityActor(SingletonActor, SingletonEntityId, improbable::CreateOffsetMapFromActor(SingletonActor, Info));
+
+	UE_LOG(LogGlobalStateManager, Log, TEXT("Linked Singleton Actor %s with id %d"), *SingletonActor->GetClass()->GetName(), SingletonEntityId);
 }
 
-void UGlobalStateManager::ExecuteInitialSingletonActorReplication()
+void UGlobalStateManager::LinkAllExistingSingletonActors()
 {
-	for (const auto& Pair : SingletonNameToEntityId)
+	// Client receive Singleton Actors via the normal Unreal replicated actor flow
+	if (!NetDriver->IsServer())
 	{
-		Worker_EntityId SingletonEntityId = Pair.Value;
+		return;
+	}
 
-		// Entity has already been created on another server
-		if (SingletonEntityId != SpatialConstants::INVALID_ENTITY_ID)
+	for (const auto& Pair : SingletonNameToEntityId)
+	{
+		UClass* SingletonActorClass = LoadObject<UClass>(nullptr, *Pair.Key);
+		if (SingletonActorClass == nullptr)
 		{
+			UE_LOG(LogGlobalStateManager, Error, TEXT("Failed to find Singleton Actor Class."));
 			continue;
 		}
 
-		AActor* SingletonActor = nullptr;
-		USpatialActorChannel* Channel = nullptr;
-		GetSingletonActorAndChannel(Pair.Key, SingletonActor, Channel);
+		LinkExistingSingletonActor(SingletonActorClass);
+	}
+}
 
-		// Class couldn't be found or channel already exists
-		if (Channel == nullptr || Channel->Actor != nullptr)
-		{
-			check(Channel == nullptr || Channel->Actor == SingletonActor);
-			continue;
-		}
+USpatialActorChannel* UGlobalStateManager::AddSingleton(AActor* SingletonActor)
+{
+	check(SingletonActor->GetIsReplicated());
 
-		if (!SingletonActor->GetIsReplicated())
-		{
-			UE_LOG(LogGlobalStateManager, Warning, TEXT("Singleton Actor %s isn't replicated! No entity will be created."), *SingletonActor->GetClass()->GetName());
-			continue;
-		}
+	UClass* SingletonActorClass = SingletonActor->GetClass();
+
+	auto& ActorChannelPair = NetDriver->SingletonActorChannels.FindOrAdd(SingletonActorClass);
```

Comment:
auto????????????????????????????????????????????
???????????????????????????????????????????????????????????????????????

Reply:
Damn you Sahil